### PR TITLE
Merge Experimental to Master

### PR DESCRIPTION
**Pull request by Ampferl**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,8 +1,148 @@
...
+    "first_name": "String",
+    "last_name": "String",
+    "teacher_token": "String"
+}
+```
+Response:\
+__Success:__\
+Status Code: `201`\
+__Error:__
+```json
```
Comment (JannikEmmerich) [inline README.md]: Please use `jsonc` instead of `json`

File: README.md [hunk 2]
```diff
@@ -1,8 +1,148 @@
...
+{
+    "token": "String"
+}
+```
+Response:\
+__Success:__\
+Status Code: `200`\
+__Error:__
+```json
```
Comment (JannikEmmerich) [inline README.md]: Please use `jsonc` instead of `json`

File: README.md [hunk 3]
```diff
@@ -1,8 +1,148 @@
...
+{
+    "email": "String",
+    "password": "String"
+}
+```
+Response:\
+__Success:__\
+Status Code: `200`\
+__Error:__
+```json
```
Comment (JannikEmmerich) [inline README.md]: Please use `jsonc` instead of `json`

File: README.md [hunk 4]
```diff
@@ -1,8 +1,148 @@
...
+{
+    "teacher": "Boolean",
+    "admin": "Boolean",
+    "privileged_student": "Boolean",
+    "report_spammer": "Number"
+}
+```
+__Error:__
+```json
```
Comment (JannikEmmerich) [inline README.md]: Please use `jsonc` instead of `json`

File: README.md [hunk 5]
```diff
@@ -1,8 +1,148 @@
...
+    "admin": "Boolean|null",
+    "privileged_student": "Boolean|null",
+    "report_spammer": "Boolean|null"
+}
+```
+Response:\
+__Success:__\
+Status Code: `200`\
+__Error:__
+```json
```
Comment (JannikEmmerich) [inline README.md]: Please use `jsonc` instead of `json`

File: README.md [hunk 6]
```diff
@@ -1,8 +1,148 @@
...
+{
+    "session_token": "string"
+}
+```
+Response:\
+__Success:__\
+Status Code: `200`\
+__Error:__
+```json
```
Comment (JannikEmmerich) [inline README.md]: Please use `jsonc` instead of `json`

File: docker-compose.yml [hunk 7]
```diff
@@ -1,3 +1,18 @@
...
+
+
+  db:
+    image: mariadb:latest
+    restart: always
+    environment:
+      MYSQL_DATABASE: userms
+      MYSQL_USER: userms
+      MYSQL_PASSWORD: userms
+      MYSQL_RANDOM_ROOT_PASSWORD: test
```
Comment (JannikEmmerich) [inline docker-compose.yml]: Please use `true` instead of `test`

File: src/main.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    return 'and update responses here'
+
+
+@app.route('/user/session', methods=['POST'])
+def user_session():
+    return 'or some session checks there'
+
+
+if __name__ == '__main__':
+    app.run(port=4450, threaded=True)
```
Comment (JannikEmmerich) [inline src/main.py]: Please use port `5000` and WSGI (Web Server Gateway Interface) `gunicorn`

File: src/user.py [hunk 9]
```diff
@@ -0,0 +1,9 @@
+class User:
+    def __init__(self, email, password, first_name, last_name, status, created, role):
```
Comment (JannikEmmerich) [inline src/user.py]: Please use an `UUID` to identify the user

File: database/Database.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    REPORTED = 2
+    BANNED = 3
+
+class Database():
+
+    #def __init__(self, host, port, username, password, database):
+    #    self.host, self.port, self.username, self.password, self.database = host, port, username, password, database
+    #    self.connection, self.cursor = None, None
+
+    def __int__(self):
```
Comment (felbinger) [inline database/Database.py]: `__init__`?

File: database/Database.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
+import mysql.connector as mariadb
+from enum import Enum
+import os
+
+class Role(Enum):
+    USER = 0
+    TEACHER = 1
+    ADMIN = 2
+
+class AccountStatus(Enum):
```
Comment (felbinger) [inline database/Database.py]: import AccountStatus from `database/accountStatus.py`

File: database/database.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    def connect(self):
+        try:
+            self.connection = mariadb.connect(host=self.host, port=str(self.port), user=self.username, password=self.password, database=self.database)
+            self.cursor = self.connection.cursor(prepared=True)
+            self.connection.commit()
+            print("[DATABASE] Database connected to '" + self.host + ":" + str(self.port) + "'")
+        except: print("[DATABASE] Error: Failed to connect!")
+        
+    def setup(self):
+        self.execute('CREATE TABLE IF NOT EXISTS User_Users (userId INT NOT NULL AUTO_INCREMENT, firstName VARCHAR(255), lastName VARCHAR(255), email VARCHAR(255), password VARCHAR(255), accountStatus INT, createdAt TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP(), role INT, PRIMARY KEY (userId))')
```
Comment (JannikEmmerich) [inline database/database.py]: Please use an `UUID` as userId

File: database/database.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+    def existUser(self, email):
+        self.checkConnection()
+        if not self.isConnected(): return None
+        self.execute('SELECT userId FROM User_Users WHERE email = ?', (email,))
+        return self.cursor.fetchone() is not None
+
+    def createUser(self, firstName, lastName, email, hashedPassword, role: role.Role):
+        self.checkConnection()
+        if not self.isConnected(): return None
+        self.execute('INSERT INTO User_Users (firstName, lastName, email, password, accountStatus, role) VALUES (?, ?, ?, ?, ?, ?)', (firstName, lastName, email, hashedPassword, 0, role.value))
```
Comment (JannikEmmerich) [inline database/database.py]: Please use an `UUID` as userId

File: src/main.py [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
```
Comment (felbinger) [inline src/main.py]: Change back to default port (`5000`) / remove port argument, please.

File: src/main.py [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+database.setup()
+
+@app.route('/')
+def main():
+    return 'Team User-Microservice'
+
+
+@app.route('/user/register', methods=['POST'])
+def user_register():
+    global content
```
Comment (JannikEmmerich) [inline src/main.py]: Don't use global variables here

**Review by JannikEmmerich — CHANGES_REQUESTED** [review]
Please remove ` src/__pycache__/errorResponse.cpython-37.pyc`
Please remove `database/Database.py`

ORM (SQLAlchemy)?

File: docker-compose.yml [hunk 16]
```diff
@@ -1,3 +1,18 @@
...
+    image: mariadb:latest
+    restart: always
+    environment:
+      MYSQL_DATABASE: userms
+      MYSQL_USER: userms
+      MYSQL_PASSWORD: userms
+      MYSQL_RANDOM_ROOT_PASSWORD: true
+    networks:
+      - userms
+
```
Comment (TheCataliasTNT2k) [inline docker-compose.yml:16]: include your microservice?

**Review by MarcelCoding — CHANGES_REQUESTED** [review]
Use `UUID` as `userId`!

**Review by TheCataliasTNT2k — CHANGES_REQUESTED** [review]

File: run.py [hunk 17]
```diff
@@ -0,0 +1,2 @@
+from service import app
```
Comment (TheCataliasTNT2k) [inline run.py:1]: gunicorn? https://github.com/E-Edu/user/pull/23#discussion_r395990191

File: resources/setup.sql [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+    lastName VARCHAR(255),
+    status INT,
+    role INT,
+    createdAt TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP(),
+    PRIMARY KEY (uuid)
+);
+
+CREATE TABLE IF NOT EXISTS
+User_Teachers (
+    user_uuid VARCHAR(64) NOT NULL,
```
Comment (TheCataliasTNT2k) [inline resources/setup.sql:16]: use foreign keys

**Review by TheCataliasTNT2k — CHANGES_REQUESTED** [review]

**Review by JannikEmmerich — APPROVED** [review]

**Pull request merged**